### ISSUE TITLE
【BUG】Test failed with accelerate version  0.21.0.dev0 on single machine with 8x A100 GPU

### ISSUE THREAD

**Issue by statelesshz** [huggingface/accelerate#1722]
### System Info

```Shell
- `Accelerate` version: 0.21.0.dev0
- Platform: Linux-4.15.0-184-generic-x86_64-with-glibc2.31
- Python version: 3.9.12
- Numpy version: 1.21.5
- PyTorch version (GPU?): 1.11.0+cu113 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 754.43 GB
- GPU type: NVIDIA A100-PCIE-40GB
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Install the accelerate of the main branch
2. execute `accelerate test` with 8xA100, test failed.

The output log is as follows.
```test
Running:  accelerate-launch /home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 6
stdout: Local process index: 6
stdout: Device: cuda:6
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 1
stdout: Local process index: 1
stdout: Device: cuda:1
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 7
stdout: Local process index: 7
stdout: Device: cuda:7
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 4
stdout: Local process index: 4
stdout: Device: cuda:4
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 2
stdout: Local process index: 2
stdout: Device: cuda:2
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 3
stdout: Local process index: 3
stdout: Device: cuda:3
stdout:
stdout: Mixed precision type: no
stdout:
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 5
stdout: Local process index: 5
stdout: Device: cuda:5
stdout:
stdout: Mixed precision type: no
stdout:
stdout: **Initialization**
stdout: Testing, testing. 1, 2, 3.
stdout: Distributed environment: MULTI_GPU  Backend: nccl
stdout: Num processes: 8
stdout: Process index: 0
stdout: Local process index: 0
stdout: Device: cuda:0
stdout:
stdout: Mixed precision type: no
stdout:
stdout:
stdout: **Test process execution**
stdout:
stdout: **Test split between processes as a list**
stderr: Traceback (most recent call last):
stderr:   File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 592, in <module>
stderr:     main()
stderr:   File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 562, in main
stderr:     test_split_between_processes_list()
stderr:   File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 484, in test_split_between_processes_list
stderr:     with state.split_between_processes(data, apply_padding=True) as results:
stderr:   File "/root/anaconda3/envs/wzw/lib/python3.9/contextlib.py", line 119, in __enter__
stderr:     return next(self.gen)
stderr:   File "/home/huggingface/accelerate/src/accelerate/state.py", line 894, in split_between_processes
stderr:     with PartialState().split_between_processes(inputs, apply_padding=apply_padding) as inputs:
stderr:   File "/root/anaconda3/envs/wzw/lib/python3.9/contextlib.py", line 119, in __enter__
stderr:     return next(self.gen)
stderr:   File "/home/huggingface/accelerate/src/accelerate/state.py", line 455, in split_between_processes
stderr:     yield _split_values(inputs, start_index, end_index)
stderr:   File "/home/huggingface/accelerate/src/accelerate/state.py", line 446, in _split_values
stderr:     result += [result[-1]] * (num_samples_per_process - len(result))
stderr: IndexError: list index out of range
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38933 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38934 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38935 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38936 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38937 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38938 closing signal SIGTERM
stderr: WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38939 closing signal SIGTERM
stderr: ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 7 (pid: 38940) of binary: /root/anaconda3/envs/wzw/bin/python
stderr: Traceback (most recent call last):
stderr:   File "/root/anaconda3/envs/wzw/bin/accelerate-launch", line 33, in <module>
stderr:     sys.exit(load_entry_point('accelerate', 'console_scripts', 'accelerate-launch')())
stderr:   File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 985, in main
stderr:     launch_command(args)
stderr:   File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 970, in launch_command
stderr:     multi_gpu_launcher(args)
stderr:   File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 646, in multi_gpu_launcher
stderr:     distrib_run.run(args)
stderr:   File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/run.py", line 715, in run
stderr:     elastic_launch(
stderr:   File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 131, in __call__
stderr:     return launch_agent(self._config, self._entrypoint, list(args))
stderr:   File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 245, in launch_agent
stderr:     raise ChildFailedError(
stderr: torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
stderr: ============================================================
stderr: /home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py FAILED
stderr: ------------------------------------------------------------
stderr: Failures:
stderr:   <NO_OTHER_FAILURES>
stderr: ------------------------------------------------------------
stderr: Root Cause (first observed failure):
stderr: [0]:
stderr:   time      : 2023-07-13_07:37:32
stderr:   host      : localhost
stderr:   rank      : 7 (local_rank: 7)
stderr:   exitcode  : 1 (pid: 38940)
stderr:   error_file: <N/A>
stderr:   traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
stderr: ============================================================
Traceback (most recent call last):
  File "/root/anaconda3/envs/wzw/bin/accelerate", line 33, in <module>
    sys.exit(load_entry_point('accelerate', 'console_scripts', 'accelerate')())
  File "/home/huggingface/accelerate/src/accelerate/commands/accelerate_cli.py", line 45, in main
    args.func(args)
  File "/home/huggingface/accelerate/src/accelerate/commands/test.py", line 54, in test_command
    result = execute_subprocess_async(cmd, env=os.environ.copy())
  File "/home/huggingface/accelerate/src/accelerate/test_utils/testing.py", line 417, in execute_subprocess_async
    raise RuntimeError(
RuntimeError: 'accelerate-launch /home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py' failed with returncode 1

The combined stderr from workers follows:
Traceback (most recent call last):
  File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 592, in <module>
    main()
  File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 562, in main
    test_split_between_processes_list()
  File "/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py", line 484, in test_split_between_processes_list
    with state.split_between_processes(data, apply_padding=True) as results:
  File "/root/anaconda3/envs/wzw/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/home/huggingface/accelerate/src/accelerate/state.py", line 894, in split_between_processes
    with PartialState().split_between_processes(inputs, apply_padding=apply_padding) as inputs:
  File "/root/anaconda3/envs/wzw/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/home/huggingface/accelerate/src/accelerate/state.py", line 455, in split_between_processes
    yield _split_values(inputs, start_index, end_index)
  File "/home/huggingface/accelerate/src/accelerate/state.py", line 446, in _split_values
    result += [result[-1]] * (num_samples_per_process - len(result))
IndexError: list index out of range
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38933 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38934 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38935 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38936 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38937 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38938 closing signal SIGTERM
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 38939 closing signal SIGTERM
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 7 (pid: 38940) of binary: /root/anaconda3/envs/wzw/bin/python
Traceback (most recent call last):
  File "/root/anaconda3/envs/wzw/bin/accelerate-launch", line 33, in <module>
    sys.exit(load_entry_point('accelerate', 'console_scripts', 'accelerate-launch')())
  File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 985, in main
    launch_command(args)
  File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 970, in launch_command
    multi_gpu_launcher(args)
  File "/home/huggingface/accelerate/src/accelerate/commands/launch.py", line 646, in multi_gpu_launcher
    distrib_run.run(args)
  File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/run.py", line 715, in run
    elastic_launch(
  File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 131, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/root/anaconda3/envs/wzw/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 245, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
/home/huggingface/accelerate/src/accelerate/test_utils/scripts/test_script.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2023-07-13_07:37:32
  host      : localhost
  rank      : 7 (local_rank: 7)
  exitcode  : 1 (pid: 38940)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================

```

### Expected behavior

`accelerate test` OK with 8x A100 GPU

**Comment by statelesshz** [comment]
```diff
def test_split_between_processes_list():
    state = AcceleratorState()
    data = list(range(0, 2 * state.num_processes))  # 0 - 15
    with state.split_between_processes(data) as results:
        assert (
            len(results) == 2
        ), f"Each process did not have two items. Process index: {state.process_index}; Length: {len(results)}"

    -data = list(range(0, (2 * state.num_processes) + 3))
    +data = list(range(0, (3 * state.num_processes) - 1))

    with state.split_between_processes(data, apply_padding=True) as results:
        if state.is_last_process:
            # Test that the last process gets the extra item(s)
            num_samples_per_device = math.ceil(len(data) / state.num_processes)
            assert (
                len(results) == num_samples_per_device
            ), f"Last process did not get the extra item(s). Process index: {state.process_index}; Length: {len(results)}"
    state.wait_for_everyone()
```

**Comment by sgugger** [comment]
cc @muellerzr

**Comment by muellerzr** [comment]
Closed now due to #1724
